File: scripts/security-advisory-automation/src/advisory_automation/branches.py
from __future__ import annotations

import re

# A platform-version composer.json *conflict* bound: "<low || >=high".
# The compatible line is the start of the gap (the value after "<").
_CONFLICT_RE = re.compile(r"<\s*([0-9][0-9.]*)\s*\|\|\s*>=\s*([0-9][0-9.]*)")


def is_unified_era(platform_line: str) -> bool:
    """2026.1 and later use unified branch naming (repo version == platform version)."""
    return int(platform_line.split(".")[0]) >= 2026


def parse_compatible_line(conflict_constraint: str) -> str:
    """Return the compatible version line from a conflict bound like '<2.3 || >=3'."""
    match = _CONFLICT_RE.search(conflict_constraint or "")
    if not match:
        raise ValueError(f"unrecognised conflict constraint: {conflict_constraint!r}")
    return match.group(1)


def _version_key(line: str) -> tuple[int, ...]:
    return tuple(int(part) for part in line.split("."))


def select_lowest_active_line(
    candidate_lines: list[str], eol_lines: set[str]
) -> str:
    """Lowest non-EOL bugfix line by numeric version order."""
    active = [line for line in candidate_lines if line not in set(eol_lines)]
    if not active:
        raise ValueError("no active bugfix lines among candidates")
    return min(active, key=_version_key)


def ee_repo_name(base_repo: str) -> str:
    """Derive the conventional ee-* LTS counterpart: 'owner/name' -> 'owner/ee-name'."""
    if "/" not in base_repo:
        raise ValueError(f"expected 'owner/name', got {base_repo!r}")
    owner, name = base_repo.split("/", 1)
    return f"{owner}/ee-{name}"


def select_branch_repo(
    base_repo: str, branch_in_base: bool, branch_in_ee: bool
) -> str:
    """Where the target branch lives: base repo first, else the conventional
    ee-* counterpart. Raises if neither has it (human-fallback — this also
    covers packages that simply have no ee-* repo).

    Branch-presence is supplied by the caller's I/O (Plan 2): branch_in_ee is
    True only when the derived ee-* repo exists AND has the branch.
    """
    if branch_in_base:
        return base_repo
    if branch_in_ee:
        return ee_repo_name(base_repo)
    raise ValueError(
        f"branch not found in base repo {base_repo!r} or its ee-* counterpart"
    )

File: scripts/security-advisory-automation/lib/branches.js
'use strict';

// A platform-version composer.json conflict bound: "<low || >=high".
// The compatible line is the start of the gap (the value after "<").
const _CONFLICT_RE = /<\s*([0-9][0-9.]*)\s*\|\|\s*>=\s*([0-9][0-9.]*)/;

/**
 * 2026.1 and later use unified branch naming (repo version == platform version).
 * @param {string} line
 * @returns {boolean}
 */
function isUnifiedEra(line) {
  return parseInt(line.split('.')[0], 10) >= 2026;
}

/**
 * Return the compatible version line from a conflict bound like '<2.3 || >=3'.
 * @param {string} constraint
 * @returns {string}
 */
function parseCompatibleLine(constraint) {
  const match = _CONFLICT_RE.exec(constraint || '');
  if (!match) {
    throw new Error(`unrecognised conflict constraint: ${JSON.stringify(constraint)}`);
  }
  return match[1];
}

/**
 * @param {string} line
 * @returns {number[]}
 */
function _versionKey(line) {
  return line.split('.').map(p => parseInt(p, 10));
}

/**
 * Compare two numeric version tuples element-wise (e.g. [2,9] < [11,0]).
 * @param {number[]} a
 * @param {number[]} b
 * @returns {number}
 */
function _compareVersions(a, b) {
  const len = Math.max(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const av = a[i] || 0;
    const bv = b[i] || 0;
    if (av !== bv) return av - bv;
  }
  return 0;
}

/**
 * Lowest non-EOL bugfix line by numeric version order.
 * @param {string[]} candidateLines
 * @param {string[]|Set<string>} eolLines
 * @returns {string}
 */
function selectLowestActiveLine(candidateLines, eolLines) {
  const eolSet = eolLines instanceof Set ? eolLines : new Set(eolLines);
  const active = candidateLines.filter(line => !eolSet.has(line));
  if (active.length === 0) {
    throw new Error('no active bugfix lines among candidates');
  }
  return active.reduce((lowest, line) =>
    _compareVersions(_versionKey(line), _versionKey(lowest)) < 0 ? line : lowest
  );
}

/**
 * Derive the conventional ee-* LTS counterpart: 'owner/name' -> 'owner/ee-name'.
 * @param {string} baseRepo
 * @returns {string}
 */
function eeRepoName(baseRepo) {
  if (!baseRepo.includes('/')) {
    throw new Error(`expected 'owner/name', got ${JSON.stringify(baseRepo)}`);
  }
  const slash = baseRepo.indexOf('/');
  const owner = baseRepo.slice(0, slash);
  const name = baseRepo.slice(slash + 1);
  return `${owner}/ee-${name}`;
}

/**
 * Where the target branch lives: base repo first, else the conventional
 * ee-* counterpart. Throws if neither has it.
 * @param {string} baseRepo
 * @param {boolean} branchInBase
 * @param {boolean} branchInEe
 * @returns {string}
 */
function selectBranchRepo(baseRepo, branchInBase, branchInEe) {
  if (branchInBase) return baseRepo;
  if (branchInEe) return eeRepoName(baseRepo);
  throw new Error(
    `branch not found in base repo ${JSON.stringify(baseRepo)} or its ee-* counterpart`
  );
}

module.exports = {
  isUnifiedEra,
  parseCompatibleLine,
  selectLowestActiveLine,
  eeRepoName,
  selectBranchRepo,
};
